File: backend/src/main/java/com/expensesage/service/ExpenseService.java
package com.expensesage.service;

import java.util.List;

import com.expensesage.dto.ExpenseCreateRequest;
import com.expensesage.model.Expense;
import com.expensesage.model.User;

public interface ExpenseService {

    /**
     * Creates a new expense, including payer details, and calculates the corresponding splits.
     *
     * @param request     The DTO containing expense details, payer details, and split information.
     * @param currentUser The user creating the expense (used for permission checks if needed).
     * @return The newly created Expense entity, including its payers and calculated splits.
     * @throws RuntimeException if validation fails (e.g., payer/split amounts don't sum correctly, invalid users/group).
     */
    Expense createExpense(ExpenseCreateRequest request, org.springframework.web.multipart.MultipartFile receiptFile, User currentUser); // Added receiptFile param

    /**
     * Retrieves an expense by its ID, ensuring the current user has access

    /**
     * Retrieves an expense by its ID, ensuring the current user has access
     * (e.g., they paid, are owed, or are in the associated group).
     *
     * @param expenseId   The ID of the expense.
     * @param currentUser The user requesting the expense details.
     * @return The Expense entity.
     * @throws RuntimeException if expense not found or user lacks access.
     */
    Expense getExpenseById(Long expenseId, User currentUser);

    /**
     * Retrieves all expenses related to the current user (paid by them or owed by
     * them).
     * This might involve querying both Expenses and Splits.
     * (Further refinement needed for optimal implementation).
     *
     * @param currentUser The user whose expenses to retrieve.
     * @return A list of relevant Expense entities.
     */
    List<Expense> getExpensesForUser(User currentUser);

    /**
     * Retrieves all expenses associated with a specific group.
     *
     * @param groupId     The ID of the group.
     * @param currentUser The user requesting the list (must be a member of the
     *                    group).
     * @return A list of Expense entities belonging to the group.
     * @throws RuntimeException if group not found or user is not a member.
     */
    List<Expense> getExpensesForGroup(Long groupId, User currentUser);

    /**
     * Updates an existing expense. Only the payer can update for now.
     * Recalculates splits based on the updated request.
     *
     * @param expenseId   The ID of the expense to update.
     * @param request     The DTO containing updated expense details and split
     *                    information.
     * @param currentUser The user attempting the update.
     * @return The updated Expense entity.
     * @throws RuntimeException if expense not found, validation fails, or
     *                          permission denied.
     */
    Expense updateExpense(Long expenseId, ExpenseCreateRequest request, org.springframework.web.multipart.MultipartFile receiptFile, User currentUser); // Added receiptFile param

    /**
     * Deletes an expense. Only the payer can delete for now.
     * Consider implications if payments have already been made based on this
     * expense.
     *
     * @param expenseId   The ID of the expense to delete.
     * @param currentUser The user attempting the deletion.
     * @throws RuntimeException if expense not found or permission denied.
     */
    void deleteExpense(Long expenseId, User currentUser);

}

File: frontend/src/services/expenseService.ts
import apiClient from './api';
import { ExpenseCreateRequest, ExpenseResponseDto } from '../types/api';

// Fetches all expenses involving the current user
export const getMyExpenses = async (): Promise<ExpenseResponseDto[]> => {
  try {
    const response = await apiClient.get<ExpenseResponseDto[]>('/expenses/my');
    return response.data;
  } catch (error: any) {
    throw error.response?.data || new Error('Failed to fetch expenses');
  }
};

// Creates a new expense
export const createExpense = async (expenseData: ExpenseCreateRequest, receiptFile?: File | null): Promise<ExpenseResponseDto> => {
    const formData = new FormData();
    formData.append('expenseData', JSON.stringify(expenseData));
    if (receiptFile) {
        formData.append('receiptFile', receiptFile);
    }

    try {
        const response = await apiClient.post<ExpenseResponseDto>('/expenses', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return response.data;
    } catch (error: any) {
        throw error.response?.data || new Error('Failed to create expense');
    }
};

// Updates an existing expense
export const updateExpense = async (expenseId: number, expenseData: ExpenseCreateRequest, receiptFile?: File | null): Promise<ExpenseResponseDto> => {
    const formData = new FormData();
    const formattedData = {
        ...expenseData,
        date: expenseData.date ? new Date(expenseData.date).toISOString().split('T')[0] : new Date().toISOString().split('T')[0],
    };
    formData.append('expenseData', JSON.stringify(formattedData));
    if (receiptFile) {
        formData.append('receiptFile', receiptFile);
    }

    try {
        const response = await apiClient.put<ExpenseResponseDto>(`/expenses/${expenseId}`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return response.data;
    } catch (error: any) {
        throw error.response?.data || new Error(`Failed to update expense ${expenseId}`);
    }
};


// Deletes an expense
export const deleteExpense = async (expenseId: number): Promise<void> => {
    try {
        await apiClient.delete(`/expenses/${expenseId}`);
    } catch (error: any) {
        throw error.response?.data || new Error(`Failed to delete expense ${expenseId}`);
    }
};